### PR TITLE
Support aliases for NodePort services

### LINKED ISSUE #718: Feature request: Add DNS A records of nodes in Route53 via dns-controller
### Use case

If we're running a UDP service (like StatsD), and we need to expose it publicly, we can use only `NodePorts` since `ELBs` don't work with UDP.
### Solution

One possible solution is keeping updated A records for all the nodes in Route53, and using a pre-defined `nodePort` value.

So, it would be awesome if `dns-controller` can create and keep the records up-to-date with the IPs of healthy nodes. Extra options like setting a TTL value would also be nice to have.

cc: @justinsb

### PR DESCRIPTION
**Pull request by justinsb**
Fixes #718

### CODE REVIEW

**Comment by justinsb** [comment]
Was able to test this for real ... made one fix, but seems to work great :-)

File: dns-controller/pkg/watchers/servicecontroller.go [hunk 1]
```diff
@@ -141,21 +141,34 @@ func (c *ServiceController) updateServiceRecords(service *v1.Service) {
...
-				ingresses = append(ingresses, dns.Record{
-					RecordType: dns.RecordTypeA,
-					Value:      ingress.IP,
-				})
+
+		if service.Spec.Type == v1.ServiceTypeLoadBalancer {
+			for i := range service.Status.LoadBalancer.Ingress {
+				ingress := &service.Status.LoadBalancer.Ingress[i]
+				if ingress.Hostname != "" {
+					// TODO: Support ELB aliases
```
Comment (kris-nova) [inline dns-controller/pkg/watchers/servicecontroller.go]: Do we need an issue here? How big is the scope on supporting these aliases?

File: dns-controller/pkg/watchers/servicecontroller.go [hunk 2]
```diff
@@ -141,21 +141,34 @@ func (c *ServiceController) updateServiceRecords(service *v1.Service) {
...
+						Value:      ingress.Hostname,
+					})
+					glog.V(4).Infof("Found CNAME for service %s/%s: %q", service.Namespace, service.Name, ingress.Hostname)
+				}
+				if ingress.IP != "" {
+					ingresses = append(ingresses, dns.Record{
+						RecordType: dns.RecordTypeA,
+						Value:      ingress.IP,
+					})
+					glog.V(4).Infof("Found A for service %s/%s: %q", service.Namespace, service.Name, ingress.IP)
```
Comment (kris-nova) [inline dns-controller/pkg/watchers/servicecontroller.go]: Do we want `A` or `A record` here?

File: dns-controller/pkg/watchers/servicecontroller.go [hunk 3]
```diff
@@ -141,21 +141,34 @@ func (c *ServiceController) updateServiceRecords(service *v1.Service) {
...
+				}
 			}
+		} else if service.Spec.Type == v1.ServiceTypeNodePort {
+			ingresses = append(ingresses, dns.Record{
+				RecordType: dns.RecordTypeAlias,
+				Value:      dns.AliasForNodesInRole("node"),
+			})
+			glog.V(4).Infof("Setting internal alias for NodePort service %s/%s", service.Namespace, service.Name)
+		} else {
+			glog.V(2).Infof("Cannot expose service %s/%s of type %q", service.Namespace, service.Name, service.Spec.Type)
```
Comment (kris-nova) [inline dns-controller/pkg/watchers/servicecontroller.go]: Looks like this function doesn't support returning `errors`.. Do we need that here? Or is a log message sufficient when we can't expose a service?

If this is fine - Should we bump the message higher than `V(2)`?

File: dns-controller/pkg/watchers/servicecontroller.go [hunk 4]
```diff
@@ -171,7 +184,7 @@ func (c *ServiceController) updateServiceRecords(service *v1.Service) {
 			}
 		}
 	} else {
-		glog.V(4).Infof("Service %q did not have %s annotation", service.Name, AnnotationNameDnsInternal)
+		glog.V(4).Infof("Service %s/%s did not have %s annotation", service.Namespace, service.Name, AnnotationNameDnsExternal)
```
Comment (kris-nova) [inline dns-controller/pkg/watchers/servicecontroller.go]: Same as above. 

Are we kosher without throwing an `error` upstream? And is `V(4)` still an okay level to keep this at?

Just smells like a failure is all - so due diligence dictates I ask.

Cheers

**Comment by ApsOps** [comment]
@justinsb Thanks for this!

I just tested - `dns.alpha.kubernetes.io/external` annotation works like a charm, but `dns.alpha.kubernetes.io/internal` does not. Is that not supported or am I missing something?

It's also reporting `node changed` continuously even if they haven't. I have 2 fixed nodes in this cluster as evident from the logs:
```
I1108 09:49:37.354102       1 dnscontroller.go:147] No changes since DNS values last successfully applied
I1108 09:49:39.509760       1 nodecontroller.go:111] node changed: MODIFIED ip-10-0-13-134.ec2.internal
I1108 09:49:39.509947       1 dnscontroller.go:476] Update node/ip-10-0-13-134.ec2.internal: [{A node/ip-10-0-13-134.ec2.internal/internal 10.0.13.134 true} {A node/ip-10-0-13-134.ec2.internal/external x.x.x.x true} {A node/role=node/external x.x.x.x true}]
I1108 09:49:39.611769       1 nodecontroller.go:111] node changed: MODIFIED ip-10-0-9-231.ec2.internal
I1108 09:49:39.611914       1 dnscontroller.go:476] Update node/ip-10-0-9-231.ec2.internal: [{A node/ip-10-0-9-231.ec2.internal/internal 10.0.9.231 true} {A node/ip-10-0-9-231.ec2.internal/external y.y.y.y true} {A node/role=node/external y.y.y.y true}]
I1108 09:49:47.378252       1 dnscontroller.go:147] No changes since DNS values last successfully applied
I1108 09:49:49.730370       1 nodecontroller.go:111] node changed: MODIFIED ip-10-0-9-231.ec2.internal
I1108 09:49:49.739132       1 dnscontroller.go:476] Update node/ip-10-0-9-231.ec2.internal: [{A node/ip-10-0-9-231.ec2.internal/internal 10.0.9.231 true} {A node/ip-10-0-9-231.ec2.internal/external y.y.y.y true} {A node/role=node/external y.y.y.y true}]
I1108 09:49:49.915757       1 nodecontroller.go:111] node changed: MODIFIED ip-10-0-13-134.ec2.internal
I1108 09:49:49.916357       1 dnscontroller.go:476] Update node/ip-10-0-13-134.ec2.internal: [{A node/ip-10-0-13-134.ec2.internal/internal 10.0.13.134 true} {A node/ip-10-0-13-134.ec2.internal/external x.x.x.x true} {A node/role=node/external x.x.x.x true}]
I1108 09:49:57.419739       1 dnscontroller.go:147] No changes since DNS values last successfully applied
I1108 09:49:59.856186       1 nodecontroller.go:111] node changed: MODIFIED ip-10-0-9-231.ec2.internal
I1108 09:49:59.856370       1 dnscontroller.go:476] Update node/ip-10-0-9-231.ec2.internal: [{A node/ip-10-0-9-231.ec2.internal/internal 10.0.9.231 true} {A node/ip-10-0-9-231.ec2.internal/external y.y.y.y true} {A node/role=node/external y.y.y.y true}]
```

File: dns-controller/pkg/watchers/servicecontroller.go (hunk 3 above)
Comment (justinsb) [inline dns-controller/pkg/watchers/servicecontroller.go]: It's not an error that should block the controller.  It's a user configuration error.

I think the way we normally do this is to emit an event, so it shows up in `kubectl describe service`.  I'll add a TODO

File: dns-controller/pkg/watchers/servicecontroller.go (hunk 1 above)
Comment (justinsb) [inline dns-controller/pkg/watchers/servicecontroller.go]: Opened #893

**Comment by justinsb** [comment]
@ApsOps the "node changed" is benign ... it is just reporting that it saw a change in the node, but you'll note that it detected that there was no DNS change.  When we reduce the log level these messages will go away.

What do you think an internal annotation on a service should do @ApsOps ?  Services already have a DNS name inside the cluster, so I'm not really sure what we'd gain by having another name...

**Comment by ApsOps** [comment]
@justinsb I expect internal annotation to add private node IPs in Route53, instead of public ones.

**Comment by ApsOps** [comment]
@chrislovecnm these changes are included in #922 so we should probably merge just that?

**Comment by chrislovecnm** [comment]
Well I would rather not have the changes in #922 :)  If we need to cherry pick and remove a PR...

**Comment by ApsOps** [comment]
@chrislovecnm Makes sense. I'll rebase #922 once this is merged.

**Comment by kris-nova** [comment]
This LGTM - @ApsOps Can you rebase #922 now please? Merging now.

**Comment by ApsOps** [comment]
@kris-nova Rebased #922, PTAL.

**Pull request merged**